### PR TITLE
dm: add dumpstatus in query-status

### PR DESCRIPTION
**Pull request by docsir**
### What problem does this PR solve? 
Close https://github.com/pingcap/ticdc/issues/3458

### What is changed and how it works?
DM can display progress for dump phase in query-status

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update key monitor metrics in both TiCDC document and official document

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/dumpling/dumpling.go [hunk 1]
```diff
@@ -196,7 +201,43 @@ func (m *Dumpling) Update(context.Context, *config.SubTaskConfig) error {
 // Status implements Unit.Status.
 func (m *Dumpling) Status(_ *binlog.SourceStatus) interface{} {
 	// NOTE: try to add some status, like dumped file count
-	return &pb.DumpStatus{}
+	mid := &export.Midparams{}
+	mid = m.realdumper.GetParameters()
```
Comment (lance6716) [inline dm/dumpling/dumpling.go:205]: should add a test about calling `m.realdumper.GetParameters()` after `dumpling.Dump()`

File: dm/dumpling/dumpling.go [hunk 2]
```diff
@@ -196,7 +201,43 @@ func (m *Dumpling) Update(context.Context, *config.SubTaskConfig) error {
...
+		FinishedRows:      mid.FinishedRows,
+		EstimateTotalRows: mid.EstimateTotalRows,
+	}
+	go m.printStatus()
+	return s
+}
+
+const logProgressTick = 2 * time.Minute
+
+func (m *Dumpling) printStatus() {
```
Comment (lance6716) [inline dm/dumpling/dumpling.go:219]: this function is added for other units to peroidically print status to log, but dumpling itself will do this so we don't bother write `printStatus` in DM

Comment (docsir) [reply dm/dumpling/dumpling.go:219]: emm, I understand that I should remove function `printStatus()`, but how dumpling print status by itself?

File: dm/dumpling/dumpling.go (hunk 1 above)
Comment (docsir) [reply dm/dumpling/dumpling.go:205]: I agree and I will try to add this test

File: dm/dumpling/dumpling.go (hunk 2 above)
Comment (lance6716) [reply dm/dumpling/dumpling.go:219]: https://github.com/pingcap/tidb/blob/cbe5240dd0367d8a6ae882e869dd3a340fb15e77/dumpling/export/dump.go#L223

And dumpling use the same logger with DM, so logs are written to same file.

File: dm/dm/proto/dmworker.proto [hunk 3]
```diff
@@ -101,6 +101,11 @@ message CheckStatus {
 // DumpStatus represents status for dump unit
 // add fields later
 message DumpStatus {
+    int64 totalTables = 1;
```
Comment (Ehco1996) [inline dm/dm/proto/dmworker.proto:104]: thanks for you work, could please also add this dump satus to OpenAPI ?

now we only have LoaderStatus and SycnerStatus in OpenAPI, you can do this work by following  these steps

1. add a new schema called`DumpStatus` in openapi yaml file [here](https://github.com/pingcap/ticdc/blob/3a2d2c5a5e9eec0eed946c5dfbdba3ce7210e526/dm/openapi/spec/dm.yaml#L1033)
2. use `make dm_generate_openapi` to generate stuct you added in yaml
3. add dump status in `openapi.go` like [loader/syncer](https://github.com/pingcap/ticdc/blob/3a2d2c5a5e9eec0eed946c5dfbdba3ce7210e526/dm/dm/master/openapi.go#L589) status 
4. add UT/IT for the new feature

btw you can open an issue to track this job and do this in next pr after this pr merged

File: dm/dumpling/dumpling.go [hunk 4]
```diff
@@ -46,6 +48,7 @@ type Dumpling struct {
 
 	dumpConfig *export.Config
 	closed     atomic.Bool
+	realdumper *export.Dumper
```
Comment (Ehco1996) [inline dm/dumpling/dumpling.go:51]: maybe we can call this `core` like lightning https://github.com/pingcap/ticdc/blob/3a2d2c5a5e9eec0eed946c5dfbdba3ce7210e526/dm/loader/lightning.go#L60

File: dm/dm/proto/dmworker.proto (hunk 3 above)
Comment (docsir) [reply dm/dm/proto/dmworker.proto:104]: thanks for your suggestions, I am confident that I can finish this task

File: go.sum [hunk 5]
```diff
@@ -833,8 +838,13 @@ github.com/pingcap/sysutil v0.0.0-20210315073920-cc0985d983a3/go.mod h1:tckvA041
 github.com/pingcap/sysutil v0.0.0-20210730114356-fcd8a63f68c5 h1:7rvAtZe/ZUzOKzgriNPQoBNvleJXBk4z7L3Z47+tS98=
 github.com/pingcap/sysutil v0.0.0-20210730114356-fcd8a63f68c5/go.mod h1:XsOaV712rUk63aOEKYP9PhXTIE3FMNHmC2r1wX5wElY=
 github.com/pingcap/tidb v1.1.0-beta.0.20211023132847-efa94595c071/go.mod h1:Ci7ABF58a4jn6YtaHi7655jP409edqC2JxWWFRqOubg=
+<<<<<<< HEAD
```
Comment (lance6716) [inline go.sum:841]: should correctly resolve the git conflicts, decide the content to leave in files and remove lines like `<<<<<<<` `=======` `>>>>>>>`

File: dm/dumpling/dumpling.go [hunk 6]
```diff
@@ -193,10 +199,26 @@ func (m *Dumpling) Update(context.Context, *config.SubTaskConfig) error {
 	return nil
 }
 
+var mu sync.Mutex
```
Comment (lance6716) [inline dm/dumpling/dumpling.go:202]: this variable should belongs to `m *Dumpling`

File: dm/dumpling/dumpling.go [hunk 7]
```diff
@@ -193,10 +199,26 @@ func (m *Dumpling) Update(context.Context, *config.SubTaskConfig) error {
...
+
 // Status implements Unit.Status.
 func (m *Dumpling) Status(_ *binlog.SourceStatus) interface{} {
 	// NOTE: try to add some status, like dumped file count
-	return &pb.DumpStatus{}
+	mu.Lock()
+	if m.realdumper != nil {
```
Comment (lance6716) [inline dm/dumpling/dumpling.go:208]:
```suggestion
	if m.realdumper == nil {
```

File: dm/dumpling/dumpling_test.go [hunk 8]
```diff
@@ -124,3 +126,26 @@ func (d *testDumplingSuite) TestDefaultConfig(c *C) {
 	c.Assert(dumpling.dumpConfig.StatementSize, Not(Equals), export.UnspecifiedSize)
 	c.Assert(dumpling.dumpConfig.Rows, Not(Equals), export.UnspecifiedSize)
 }
+
+func TestCallStatus(t *testing.T) {
+	conf := export.DefaultConfig()
+	cfg := &config.SubTaskConfig{}
+	m := &Dumpling{cfg: cfg}
+	ctx := context.Background()
+
```
Comment (lance6716) [inline dm/dumpling/dumpling_test.go:135]: also check `m.Status` before `m.realdumper` is assigned

File: dm/dumpling/dumpling.go (hunk 6 above)
Comment (docsir) [reply dm/dumpling/dumpling.go:202]: I understand that this variable is one of `m *Dumpling` fields, so I should add a new field to `*Dumpling`

File: dm/dumpling/dumpling_test.go (hunk 8 above)
Comment (docsir) [reply dm/dumpling/dumpling_test.go:135]: I tried to check before `m.realdumper`, but it always failed, so please give me more advice

File: dm/dumpling/dumpling_test.go [hunk 9]
```diff
@@ -124,3 +125,32 @@ func (d *testDumplingSuite) TestDefaultConfig(c *C) {
 	c.Assert(dumpling.dumpConfig.StatementSize, Not(Equals), export.UnspecifiedSize)
 	c.Assert(dumpling.dumpConfig.Rows, Not(Equals), export.UnspecifiedSize)
 }
+
+func (d *testDumplingSuite) TestCallStatus(c *C) {
+	conf := export.DefaultConfig()
+	cfg := &config.SubTaskConfig{}
```
Comment (lance6716) [inline dm/dumpling/dumpling_test.go:131]: we can give these two configs a more meaningful name, such as dumpCfg, taskCfg

File: dm/dumpling/dumpling_test.go [hunk 10]
```diff
@@ -124,3 +125,32 @@ func (d *testDumplingSuite) TestDefaultConfig(c *C) {
...
+	m := &Dumpling{cfg: cfg}
+	ctx := context.Background()
+
+	ss := &binlog.SourceStatus{}
+	s := m.Status(ss).(*pb.DumpStatus)
```
Comment (lance6716) [inline dm/dumpling/dumpling_test.go:136]: we can use `m.Status(nil).(*pb.DumpStatus)` so no need `ss := &binlog.SourceStatus{}`

File: dm/dumpling/dumpling_test.go [hunk 11]
```diff
@@ -124,3 +125,32 @@ func (d *testDumplingSuite) TestDefaultConfig(c *C) {
...
+	c.Assert(s.CompletedTables, Equals, float64(0))
+	c.Assert(s.FinishedBytes, Equals, float64(0))
+	c.Assert(s.FinishedRows, Equals, float64(0))
+	c.Assert(s.EstimateTotalRows, Equals, float64(0))
+
+	dumpling, err := export.NewDumper(ctx, conf)
+	if err != nil {
```
Comment (lance6716) [inline dm/dumpling/dumpling_test.go:143]: we should use `c.Assert` to check err is nil, then assign `m.realdumper`

BTW, I also like the name `m.core` as recommended in https://github.com/pingcap/ticdc/pull/3694/files#r761572067

File: dm/dumpling/dumpling_test.go [hunk 12]
```diff
@@ -124,3 +125,32 @@ func (d *testDumplingSuite) TestDefaultConfig(c *C) {
...
+		m.realdumper = dumpling
+	}
+
+	export.InitMetricsVector(conf.Labels)
+
+	m.Close()
+	ss = &binlog.SourceStatus{}
+	s = m.Status(ss).(*pb.DumpStatus)
```
Comment (lance6716) [inline dm/dumpling/dumpling_test.go:151]:
```suggestion
	s = m.Status(nil).(*pb.DumpStatus)
```

**Review by lance6716 — APPROVED** [review]

File: dm/dumpling/dumpling_test.go [hunk 13]
```diff
@@ -124,3 +124,30 @@ func (d *testDumplingSuite) TestDefaultConfig(c *C) {
...
+	ctx := context.Background()
+
+	s := m.Status(nil).(*pb.DumpStatus)
+	c.Assert(s.CompletedTables, Equals, float64(0))
+	c.Assert(s.FinishedBytes, Equals, float64(0))
+	c.Assert(s.FinishedRows, Equals, float64(0))
+	c.Assert(s.EstimateTotalRows, Equals, float64(0))
+
+	dumpling, err := export.NewDumper(ctx, dumpConf)
+	c.Assert(err, NotNil)
```
Comment (lance6716) [inline dm/dumpling/dumpling_test.go:141]:
```suggestion
	c.Assert(err, IsNil)
```

**Review by lance6716 — CHANGES_REQUESTED** [review]

File: dm/dumpling/dumpling_test.go (hunk 13 above)
Comment (Ehco1996) [reply dm/dumpling/dumpling_test.go:141]: this error is `sql: SELECT version();: Error 1045: Access denied for user 'root'@'localhost' (using password: NO)`

i don't think our ut should access real mysql, could please try fix this error

File: dm/dumpling/dumpling_test.go [hunk 14]
```diff
@@ -124,3 +124,30 @@ func (d *testDumplingSuite) TestDefaultConfig(c *C) {
...
+	m.core = dumpling
+
+	export.InitMetricsVector(dumpConf.Labels)
+
+	m.Close()
+
+	s = m.Status(nil).(*pb.DumpStatus)
+	c.Assert(s.CompletedTables, Equals, float64(0))
```
Comment (Ehco1996) [inline dm/dumpling/dumpling_test.go:149]: can your explain the purpose of L149~L152?  what are these lines of code testing?

Comment (docsir) [reply dm/dumpling/dumpling_test.go:149]: thanks, I need more advice to fix this err. I think these lines of code can test whether the call of `func GetParameters()` is successful

File: dm/dumpling/dumpling.go [hunk 15]
```diff
@@ -196,7 +203,20 @@ func (m *Dumpling) Update(context.Context, *config.SubTaskConfig) error {
 // Status implements Unit.Status.
 func (m *Dumpling) Status(_ *binlog.SourceStatus) interface{} {
 	// NOTE: try to add some status, like dumped file count
-	return &pb.DumpStatus{}
+	m.mu.Lock()
+	defer m.mu.Unlock()
```
Comment (lichunzhu) [inline dm/dumpling/dumpling.go:207]:
```suggestion
	m.mu.RLock()
	defer m.mu.RUnlock()
```

File: dm/dumpling/dumpling_test.go (hunk 14 above)
Comment (Ehco1996) [reply dm/dumpling/dumpling_test.go:149]: for the err in L141 , you can use `conn.InitMockDB(c)` to mock db Connection, you can see a example[ here](https://github.com/pingcap/ticdc/blob/f0a38a7aaf9f3b11a4d807da275b567642733f58/dm/dumpling/util_test.go#L35)

if you want to test `GetParameters/dumpling metrics` works as expected ,i thinks add a query-status it-test  is a better choice, maybe use dmctl query-status  to check metrics not equal to zero in this [full-mode-test](https://github.com/pingcap/ticdc/blob/f0a38a7aaf9f3b11a4d807da275b567642733f58/dm/tests/full_mode/run.sh#L133)

Comment (lichunzhu) [reply dm/dumpling/dumpling_test.go:149]: Maybe you'd better use `&Dumper{}` for test.

Comment (docsir) [reply dm/dumpling/dumpling_test.go:148]: > for the err in L141 , you can use `conn.InitMockDB(c)` to mock db Connection, you can see a example[ here](https://github.com/pingcap/ticdc/blob/f0a38a7aaf9f3b11a4d807da275b567642733f58/dm/dumpling/util_test.go#L35)
> 
> if you want to test `GetParameters/dumpling metrics` works as expected ,i thinks add a query-status it-test is a better choice, maybe use dmctl query-status to check metrics not equal to zero in this [full-mode-test](https://github.com/pingcap/ticdc/blob/f0a38a7aaf9f3b11a4d807da275b567642733f58/dm/tests/full_mode/run.sh#L133)

thanks for your advice, I will add a it-tets for query-status in next PR(because I haven't run this it-test successfully)

**Review by lance6716 — APPROVED** [review]
cc @lichunzhu

**Review by lichunzhu — APPROVED** [review]

**Comment by lichunzhu** [comment]
@docsir Please help resolve the conflicts.

File: dm/dumpling/dumpling_test.go (hunk 14 above)
Comment (Ehco1996) [reply dm/dumpling/dumpling_test.go:148]: lgtm

File: dm/dumpling/dumpling_test.go [hunk 16]
```diff
@@ -129,6 +129,32 @@ func (t *testDumplingSuite) TestDefaultConfig(c *C) {
 	c.Assert(dumpling.dumpConfig.Rows, Not(Equals), export.UnspecifiedSize)
 }
 
+func (d *testDumplingSuite) TestCallStatus(c *C) {
```
Comment (lance6716) [inline dm/dumpling/dumpling_test.go:132]:
```suggestion
func (t *testDumplingSuite) TestCallStatus(c *C) {
```

CI shows a lint error

File: dm/dumpling/dumpling_test.go [hunk 17]
```diff
@@ -129,6 +129,32 @@ func (t *testDumplingSuite) TestDefaultConfig(c *C) {
 	c.Assert(dumpling.dumpConfig.Rows, Not(Equals), export.UnspecifiedSize)
 }
 
+func (t *testDumplingSuite) TestCallStatus(c *C) {
+	m := NewDumpling(d.cfg)
```
Comment (lance6716) [inline dm/dumpling/dumpling_test.go:133]:
```suggestion
	m := NewDumpling(t.cfg)
```
